src/routes.rs

```diff
@@ -526,4 +543,330 @@ pub mod data {
+                    fn id(&self) -> &str;
+                    fn insert<T: Serialize +'static>(&mut self, key: &str, value: T) -> Result<(), serde_json::Error>;
+                    fn insert_raw(&mut self, key: &str, value: String);
+                    fn get<T: serde::de::DeserializeOwned + 'static>(&self, key: &str) -> Option<T>;
+                    fn get_raw(&self, key: &str) -> Option<String>;
+                }
+            impl Clone for Session {
+                fn clone(&self) -> Self;
+            }
+                impl Debug for Session {
+                    fn fmt<'a>(&self, f: &mut Formatter<'a>) -> fmt::Result;
+                }
+            }
+
+            #[tokio::test]
+            async fn with_token_with_no_query_params() {
+                let mut session = Session::new();
+                session.set_cookie_value("testSID".to_owned());
+                session
+                    .insert(
+                        "token",
+                        "E0AE2C1C9AA2DB85DFA2FF6B4AAC7A5E51FFDAA3948BECEC353561D513E59A9C",
+                    )
+                    .unwrap();
+                let expected_sid = session.id().to_owned();
+
+                let mut mock_store = MockSessionStore::new();
+                mock_store
+                    .expect_load_session()
+                    .with(predicate::eq("testSID".to_owned()))
+                    .times(1)
+                    .return_once(move |_| Ok(Some(session)));
+                mock_store
+                    .expect_destroy_session()
+                    .withf(move |session: &Session| session.id() == expected_sid)
+                    .times(1)
+                    .return_once(move |_| Ok(()));
+                let session_store = ArcSessionStore(Arc::new(mock_store));
+
+                let mut render_engine = MockRenderer::new();
+                render_engine
+                    .expect_render()
+                    .withf(move |template: &RenderTemplate<HashMap<String, String>>| {
+                        let mut expected_value = HashMap::new();
+                        expected_value.insert("path".to_string(), ".testKey.".to_string());
+                        expected_value.insert("data".to_owned(), "testValue".to_owned());
+                        expected_value.insert("data_type".to_owned(), "string".to_owned());
+                        template.value == expected_value
+                    })
+                    .times(1)
+                    .return_once(move |_| Ok("".to_string()));
+
+                let mut token_generator = MockTokenGenerator::new();
+                token_generator
+                    .expect_generate_token()
+                    .times(1)
+                    .returning(|| {
+                        Ok(
+                            "E0AE2C1C9AA2DB85DFA2FF6B4AAC7A5E51FFDAA3948BECEC353561D513E59A9C"
+                                .to_owned(),
+                        )
+                    });
```

Comment:
I'd argue this only should be tested when the token is used (editing).  But just nitpicking.